src/app/archive-settings/manage-metadata/manage-custom-metadata/manage-custom-metadata.component.ts

```diff
@@ -0,0 +1,136 @@
+  @ViewChild('element') public element: ElementRef;
+  public categories: string[] = [];
+  public activeCategory: string | null = null;
+  public values: TagVO[] = [];
+
+  public dismissEvent = new Subject<number>();
+
+  public tagsList: TagVO[] = [];
+  protected tagMap: Map<string, TagVO[]> = new Map<string, TagVO[]>();
+
+  protected deletedTagIds: number[] = [];
+
+  protected windowListener: () => void;
+
+  constructor(private renderer: Renderer2, private tag: TagsService) {
+    this.windowListener = this.renderer.listen(
+      'window',
+      'click',
+      (e: Event) => {
+        if (e.target && !this.element.nativeElement.contains(e.target)) {
+          this.dismissEvent.next(-1);
+        }
+      }
+    );
+  }
+
+  ngOnInit(): void {
+    this.getTagsFromTagService();
+    this.tagMap.clear();
+    this.reloadTagMap();
+  }
+
+  ngOnDestroy(): void {
+    if (this.windowListener) {
+      this.windowListener();
+    }
+  }
+
+  public setActiveCategory(category: string): void {
+    if (!this.tagMap.has(category)) {
+      this.activeCategory = null;
+      this.values = [];
+      return;
+    }
+    this.activeCategory = category;
+    this.values = this.tagMap.get(category);
+  }
+
+  public async refreshTagsInPlace() {
+    await this.tag.resetTags();
+    this.getTagsFromTagService();
+    this.tagMap.clear();
+    this.reloadTagMap();
+    this.setActiveCategory(this.activeCategory);
+  }
+
+  public identifyTag(index: number, item: TagVO): number {
+    return item.tagId;
+  }
+
+  public addDeletedTag(tag: TagVO): void {
+    this.deletedTagIds.push(tag.tagId);
+    const [category, value] = tag.name.split(':');
+    if (this.tagMap.get(category).length === 1) {
+      this.activeCategory = null;
+    }
+  }
+
+  public addDeletedCategory(category: string): void {
+    if (this.tagMap.has(category)) {
+      for (const tag of this.tagMap.get(category)) {
+        this.addDeletedTag(tag);
+      }
+    }
+
+    if (this.activeCategory === category) {
+      this.activeCategory = null;
+    }
+  }
+
+  protected getTagsFromTagService() {
+    this.tagsList = this.tag
+      .getTags()
+      .map((data) => new TagVO(data))
+      .filter(
+        (tag) =>
+          tag.isCustomMetadata() && !this.deletedTagIds.includes(tag.tagId)
+      );
+  }
+
+  protected addTagToTagMap(category: string, tag?: TagVO): void {
+    if (this.tagMap.has(category)) {
+      if (tag) {
+        this.tagMap.get(category).push(tag);
+      }
+    } else {
+      if (tag) {
+        this.tagMap.set(category, [tag]);
+      } else {
+        this.tagMap.set(category, []);
+      }
+    }
+  }
+
+  protected reloadTagMap(): void {
+    for (const tag of this.tagsList) {
+      if (tag.name.indexOf(':') !== -1) {
+        const [category, value] = tag.name.split(':');
+        this.addTagToTagMap(category, tag);
+      } else {
+        this.addTagToTagMap(tag.name);
+      }
```

Comment:
I believe I put this in place in case we decide to allow creating empty categories in the future. Of course, that feature doesn't exist in the backend right now, and might work differently once implemented anyway, so this code could be removed.

Reply:
I think there's a high enough chance that we don't add that feature, or we don't add it anytime soon, that we shouldn't have code that exists solely to prepare for it; there's too much risk of that code becoming a source of confusion in the absence of the feature it's meant for.